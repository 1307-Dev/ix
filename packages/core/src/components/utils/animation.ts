/**
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */
export default abstract class Animation {
  static readonly shortTime = 0;
  static readonly defaultTime = 150;
  static readonly mediumTime = 300;
  static readonly slowTime = 500;
  static readonly xSlowTime = 1000;
}
