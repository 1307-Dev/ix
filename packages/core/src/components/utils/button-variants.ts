/*
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */
export type Buttons = 'Primary' | 'Secondary';

export enum ButtonVariants {
  'Primary' = 'btn-primary',
  'Secondary' = 'btn-secondary',
}
