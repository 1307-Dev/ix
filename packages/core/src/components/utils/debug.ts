/*
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */
import { version } from './../../env';

export function addVersion() {
  if (document?.head) {
    const commentVersion = document.createComment(
      btoa(`ix-webcomponents: ${version}`)
    );
    document.body.appendChild(commentVersion);
  }
}
