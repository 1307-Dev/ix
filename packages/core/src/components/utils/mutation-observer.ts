/*
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */
export const createMutationObserver = (callback: MutationCallback) => new MutationObserver(callback);
