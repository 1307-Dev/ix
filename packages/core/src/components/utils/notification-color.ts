// SPDX-FileCopyrightText: 2022 Siemens AG
//
// SPDX-License-Identifier: MIT

/**
 * Icon colors representing common notification states (e.g. alarm, info).
 */
export type NotificationColor = 'color-std-text' | 'color-info' | 'color-warning' | 'color-success' | 'color-alarm';
