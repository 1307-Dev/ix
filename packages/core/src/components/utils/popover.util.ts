/*
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */

import { convertToRemString } from './rwd.util';

export class Popover {
  constructor(private readonly hostElement: HTMLElement, private readonly popoverElement: HTMLElement, private readonly outsideCallback: (e: Event) => void) {}

  outside(e: Event) {
    if (!this.hostElement?.contains(e.target as HTMLElement)) {
      this.outsideCallback(e);
    }
  }

  open(maxWidth = 256) {
    this.popoverElement.style.transform = `
      translateX(${this.hostElement.offsetLeft + this.hostElement.offsetWidth}px)
      translateY(-${this.hostElement.offsetHeight}px)
    `;
    this.popoverElement.style.maxWidth = convertToRemString(maxWidth);
    document.body.addEventListener('click', this.outside.bind(this));
  }

  destroy() {
    document.body.removeEventListener('click', this.outside.bind(this));
  }
}
