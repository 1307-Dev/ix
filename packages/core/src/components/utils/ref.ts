// SPDX-FileCopyrightText: 2022 Siemens AG
//
// SPDX-License-Identifier: MIT

export interface CreateRef<T = HTMLElement> {
  (newRef: T): void;
  current: Promise<T>;
  _ref: T;
}

export function createRef<T = HTMLElement>(): CreateRef<T> {
  let ref: T;
  let resolveFn: (value: T | PromiseLike<T>) => void;

  const updateRef = (newRef: T) => {
    ref = newRef;
    resolveFn(ref);
  };

  updateRef.current = new Promise<T>(r => (resolveFn = r));

  Object.defineProperty(updateRef, 'current', {
    get: () => {
      const promise = new Promise<T>(r => (resolveFn = r));
      if (ref) {
        resolveFn(ref);
      }
      return promise;
    },
  });

  updateRef._ref = ref;
  return updateRef;
}
