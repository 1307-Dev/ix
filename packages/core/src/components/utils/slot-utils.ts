import { h } from '@stencil/core';

export function renderFormStatusSlots() {
  return [
    h('slot', { name: 'helper', slot: 'helper' }),
    h('slot', { name: 'warning', slot: 'warning' }),
    h('slot', { name: 'valid', slot: 'valid' }),
    h('slot', { name: 'invalid', slot: 'invalid' }),
    h('slot', { name: 'info', slot: 'info' })
  ];
}
  