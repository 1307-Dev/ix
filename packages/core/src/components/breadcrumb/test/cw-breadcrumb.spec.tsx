/*
 * COPYRIGHT (c) Siemens AG 2018-2022 ALL RIGHTS RESERVED.
 */

import { newSpecPage } from '@stencil/core/testing';
import { fireEvent } from '@testing-library/dom';
import { Breadcrumb } from '../breadcrumb';
import { createMutationObserver } from './../../utils/mutation-observer';

jest.mock('./../../utils/mutation-observer');
jest.mock('animejs');

describe('cw-breadcrumb', () => {
  let observerCallback: Function;
  let page: any;

  beforeEach(async () => {
    //@ts-ignore
    createMutationObserver = jest.fn(callback => {
      observerCallback = callback;
      return {
        observe: jest.fn(),
      };
    });
    page = await newSpecPage({
      components: [Breadcrumb],
      html: `
        <cw-breadcrumb>
          <cw-breadcrumb-item label="Test 1" />
          <cw-breadcrumb-item label="Test 2" />
          <cw-breadcrumb-item label="Test 3" />
          <cw-breadcrumb-item label="Test 4" />
        </cw-breadcrumb>
      `,
    });
  });

  it('renders', async () => {
    observerCallback();

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('emits an event when clicked on item', async () => {
    let callbackSpy = jest.fn();
    page.win.addEventListener('itemClick', callbackSpy);

    let item = page.doc.querySelector('[data-testid="item"]');
    fireEvent.click(item);
    await page.waitForChanges();

    expect(callbackSpy).toHaveBeenCalled();
  });
});
