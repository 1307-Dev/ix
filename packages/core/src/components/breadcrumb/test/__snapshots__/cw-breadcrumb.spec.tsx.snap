// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cw-breadcrumb renders 1`] = `
<cw-breadcrumb>
  <!---->
  <cw-dropdown></cw-dropdown>
  <div class="crumb-items">
    <div class="clickable crumb" data-breadcrumb="0" data-testid="item">
      <span class="crumb-text remove-anchor"></span>
      <span class="glyph glyph-16 glyph-chevron-right text-default-text"></span>
    </div>
    <div class="clickable crumb" data-breadcrumb="1" data-testid="item">
      <span class="crumb-text remove-anchor"></span>
      <span class="glyph glyph-16 glyph-chevron-right text-default-text"></span>
    </div>
    <div class="clickable crumb" data-breadcrumb="2" data-testid="item">
      <span class="crumb-text remove-anchor"></span>
      <span class="glyph glyph-16 glyph-chevron-right text-default-text"></span>
    </div>
    <div class="clickable crumb last remove-hover" data-breadcrumb="3" data-testid="item">
      <span class="crumb-text remove-anchor"></span>
    </div>
    <cw-breadcrumb-item label="Test 1">
      <cw-breadcrumb-item label="Test 2">
        <cw-breadcrumb-item label="Test 3">
          <cw-breadcrumb-item label="Test 4"></cw-breadcrumb-item>
        </cw-breadcrumb-item>
      </cw-breadcrumb-item>
    </cw-breadcrumb-item>
  </div>
  <cw-dropdown></cw-dropdown>
</cw-breadcrumb>
`;
